pymagicc/input.py

```diff
@@ -0,0 +1,225 @@
+        self.lines = lines
+
+    def read(self):
+        nml_end, nml_start = self._find_nml()
+
+        metadata = self.process_metadata(self.lines[nml_start:nml_end + 1])
+        metadata['header'] = "".join(self.lines[:nml_start])
+
+        # Create a stream from the remaining lines, ignoring any blank lines
+        stream = StringIO()
+        cleaned_lines = [l.strip()
+                         for l in self.lines[nml_end + 1:] if l.strip()]
+        stream.write("\n".join(cleaned_lines))
+        stream.seek(0)
+
+        df, units = self.process_data(stream, metadata)
+        metadata['units'] = units
+        return metadata, df
+
+    def _find_nml(self):
+        """
+        Find the start and end of the embedded namelist
+        :return: start, end indexs for the namelist
+        """
+        nml_start = None
+        nml_end = None
+        for i in range(len(self.lines)):
+            if self.lines[i].strip().startswith('&'):
+                nml_start = i
+
+            if self.lines[i].strip().startswith('/'):
+                nml_end = i
+        assert nml_start is not None and nml_end is not None, \
+            'Could not find namelist within {}'.format(self.fname)
+        return nml_end, nml_start
+
+    def process_metadata(self, lines):
+        # TODO: replace with f90nml.reads when released
+        parser = f90nml.Parser()
+        nml = parser._readstream(lines, {})
+        metadata = {
+            k.split('_')[1]: nml['THISFILE_SPECIFICATIONS'][k]
+            for k in nml['THISFILE_SPECIFICATIONS']
+        }
+
+        return metadata
+
+    def process_data(self, stream, metadata):
+        """
+        Extract the tabulated data from a subset of the input file
+        :param stream: A Streamlike object (nominally StringIO) containing the
+            table to be extracted
+        :param metadata: Dictionary containing
+        :return: Tuple of a pd.DataFrame containing the data and a Dict
+            containing the units for each gas present in the output. The
+            pd.DataFrame columns are named using a MultiIndex of Gas and Region
+        """
+        raise NotImplementedError()
+
+
+class MAGICC6Reader(InputReader):
+    def process_data(self, stream, metadata):
+        gas = basename(self.fname).split('_')[1]
+        df = pd.read_csv(
+            stream,
+            skip_blank_lines=True,
+            delim_whitespace=True,
+            index_col=0,
+            engine="python")
+        # Convert to a columns to a MultiIndex
+        df.columns = [
+            [gas] * len(df.columns),
+            df.columns
+        ]
+        df.index.name = 'YEAR'
+
+        units = {
+            gas: metadata['units']
+        }
+        return df, units
+
+
+class MAGICC7Reader(InputReader):
+    def process_data(self, stream, metadata):
+        gases = stream.readline().split()
+        assert gases[0] == 'GAS'
+        gases = gases[1:]
+        stream.readline()  # Unused line
```

Comment:
Not actually unused but rather only used for in certain cases. It's confusing and I'm not sure what the best thing to do with it is right now so maybe just change the comment to `TODO line, not sure how to carry this around yet so just ignoring it at the moment` and throw a warning if any value is anything other than `SET`? (I need you to explain to me how best to add commits to make these changes rather than just asking you to do them cause I'm lost at the moment).